app/models/origin.rb

```diff
@@ -0,0 +1,29 @@
+class Origin < ActiveRecord::Base
+  has_many :projects
+  has_many :contributions
+
+  validates :domain, presence: true
+  validates :domain, uniqueness: { scope: :referral }
+
+  # Process a given ref and http referrer
+  # until found or create a new Origin
+  def self.process referral, http_referrer
```

Comment:
Can't we move this logic into self.process_hash?
Or at least make this method private?